src/main/java/com/notably/logic/correction/distance/LevenshteinDistanceCalculator.java

```diff
@@ -0,0 +1,62 @@
+package com.notably.logic.correction.distance;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * An implementation of the Levenshtein algorithm for calculating the edit distance of two {@link Strings}.
+ * Inspired by https://web.stanford.edu/class/cs124/lec/med.pdf.
+ */
+public class LevenshteinDistanceCalculator implements EditDistanceCalculator {
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int calculateDistance(String first, String second) {
+        requireNonNull(first);
+        requireNonNull(second);
+
+        // Uses dynamic programming bottom-up approach in calculating the edit distance.
+        // For each pair of i,j, distance[i][j] represents the edit distance between
+        // first[0...i] and second[0...j].
+
+        int[][] distance = new int[first.length() + 1][second.length() + 1];
+
+        for (int i = 0; i <= first.length(); i++) {
+            distance[i][0] = i;
+        }
+        for (int j = 0; j <= second.length(); j++) {
+            distance[0][j] = j;
+        }
+
+        for (int i = 1; i <= first.length(); i++) {
+            for (int j = 1; j <= second.length(); j++) {
+                if (first.charAt(i - 1) == second.charAt(j - 1)) {
+                    distance[i][j] = min(
+                            distance[i - 1][j] + 1,
+                            distance[i][j - 1] + 1,
+                            distance[i - 1][j - 1]);
```

Comment:
I think if the character's match the smallest distance will definitely be distance[i - 1][j - 1], so theres no need for the check.
Otherwise, LGTM !

Reply:
Not necessarily! Take a look at page 18 of https://web.stanford.edu/class/cs124/lec/med.pdf

Reply:
Hmmm, i can't find a case in page 18 where ,distance[i - 1][j] + 1 or distance[i][j - 1] + 1 is smaller than the diagonal. It's always equal or bigger than distance[i - 1][j - 1].